### DISCUSSION TITLE
How do I serve text/html

### DISCUSSION THREAD

**Discussion by Jameswood-vega** [danielgtaylor/huma#530] [Q&A]
Hello Daniel,
we have recently adopted huma at my current employer, and have had great success with a data api we have built. 
I am trying to expand my understanding of how to use this framework, but I can't figure out how to get it to serve good old html content.  

for example I have modified the greetings example to return basic html.

I expected this to work.
```
package main

import (
	"context"
	"fmt"
	"net/http"

	"github.com/danielgtaylor/huma/v2"
	"github.com/danielgtaylor/huma/v2/adapters/humachi"
	"github.com/danielgtaylor/huma/v2/humacli"
	"github.com/go-chi/chi/v5"

	_ "github.com/danielgtaylor/huma/v2/formats/cbor"
)

// Options for the CLI. Pass `--port` or set the `SERVICE_PORT` env var.
type Options struct {
	Port int `help:"Port to listen on" short:"p" default:"8080"`
}

// GreetingOutput represents the greeting operation response.
type GreetingOutput struct {
	ContentType string `header:"Content-Type"`
	Body        struct {
		Message string `json:"message" example:"Hello, world!" doc:"Greeting message"`
	}
}

func main() {
	// Create a CLI app which takes a port option.
	cli := humacli.New(func(hooks humacli.Hooks, options *Options) {
		// Create a new router & API
		router := chi.NewMux()
		api := humachi.New(router, huma.DefaultConfig("My API", "1.0.0"))

		// Add the operation handler to the API.
		huma.Get(api, "/greeting/{name}", func(ctx context.Context, input *struct {
			Name string `path:"name" maxLength:"30" example:"world" doc:"Name to greet"`
		}) (*GreetingOutput, error) {
			resp := &GreetingOutput{}
			resp.Body.Message = fmt.Sprintf("<h1>Hello, %s!</h1>", input.Name)
			resp.ContentType = "text/html"
			return resp, nil
		})

		// Tell the CLI how to start your router.
		hooks.OnStart(func() {
			http.ListenAndServe(fmt.Sprintf(":%d", options.Port), router)
		})
	})

	// Run the CLI. When passed no commands, it starts the server.
	cli.Run()
}
```
But I get 

> 2024/07/31 16:19:14 http: panic serving 127.0.0.1:44448: error marshaling response for GET /greeting/{name} 200: unknown content type: text/html

I have tried to figure it out on my own and I am stuck please help.

**Comment by baderj** [comment]
You probably shouldn't return plain HTML at all, but you can always set the Body to `[]byte` to avoid serialization and just return raw bytes:

```golang
type GreetingOutput struct {
	ContentType string `header:"Content-Type"`
	Body        []byte `example:"<h1>Hello, world!</h1>" doc:"Greeting message"`
}
```

Then return the HTML as follows:
```golang
resp := &GreetingOutput{}
resp.Body = []byte(fmt.Sprintf("<h1>Hello, %s!</h1>", input.Name))
resp.ContentType = "text/html"
return resp, nil
```

**Reply by danielgtaylor** [reply]
Great answer, thanks! Also worth noting that if you do not need Huma's validation and input parsing then you can also just use the `router` in your example. Huma is happy to get out of the way when not needed!

```go
router.Get("/", func(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/html")
	w.Write([]byte("<html><h1>Hello, world!</h1></html>"))
})
```

**Comment by gwd** [comment]
First off, thank you for the work you've done; I'm just playing around with it and seems quite useful.

FWIW I think there are legit reasons to return pre-rendered HTML content even through an API.  My current website does HTMX-style server-side rendering of individual components of the page, which the front-end JS will request and then shove into page elements.

I managed to come up with the solution @baderj suggested, and came here to see if there was anything better.

It's OK, but it would be nicer if (e.g.) I could set `contentType:"text/html"` on the `Body` element of an output struct, and have Huma set the content type automatically.
